### ISSUE TITLE
ship stargz snapshotter

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes-sigs/kind#1981]
**What would you like to be added**:

Consider shipping stargz /crfs support in containerd by default for faster image pulls with images that support it.

https://github.com/containerd/stargz-snapshotter

**Why is this needed**:

Potential user experience improvements. Faster CI.

cc @mattmoor @AkihiroSuda

NOTE: we're probably not shipping much more in the way of features in 2020, we've lost a lot of time to bug hunting and should probably focus on wrapping up a final 2020 release. I think we should track this anyhow.

Logistically the biggest issue is probably making sure we have builds for all the architectures and that they'll work with whatever arbitrary containerd versions we are using (to keep up with various fixes, a few times we've needed mitigations for quirks / Kubernetes @ HEAD).

For containerd itself we still have https://github.com/kind-ci/containerd-nightlies to help with this (containerd *has* nightlies, but they're difficult to programmatically access / predict for install).

We also have to consider the binary size, users don't appreciate larger kind images (nor do I), so we don't typically agree to ship additional binaries that aren't strictly necessary.

**Comment by AkihiroSuda** [comment]
cc @ktock

**Comment by AkihiroSuda** [comment]
> We also have to consider the binary size, users don't appreciate larger kind images (nor do I), so we don't typically agree to ship additional binaries that aren't strictly necessary.

We can save the image size if we just add an installer script in the image.

**Comment by mattmoor** [comment]
re: image size

I did just open https://github.com/containerd/stargz-snapshotter/issues/200, one thing I noticed was that estargz doesn't expose compression level like `compress/gzip` does, and I wonder how practical that would be to add.

**Comment by mattmoor** [comment]
> We also have to consider the binary size, users don't appreciate larger kind images (nor do I), so we don't typically agree to ship additional binaries that aren't strictly necessary.

Totally ack this, but to take a devil's advocate position here: Yes, KinD is larger, but as estargz proliferates the net download users should see would be dramatically smaller.

### On a meta note

Part of why I want to push on the tooling side of this is that I don't think folks will switch unless we make it trivial or even thoughtless.  I'd like to push on this as the future default for github.com/google/go-containerregistry based tooling (ko, rules_docker, buildpacks, kaniko), but one of my struggles with that is getting mileage through it on an actual K8s cluster.

KinD has generally been my "go to" for configurable K8s for CI, so enabling folks to test this in their CI would be a benefit of earlier adoption on the curve.  I also suspect that as image builders catch up, that despite slightly more to download up-front, that their e2e testing times will improve by downloading (often much) less.


So while I don't think we should rush this (certainly as a default), providing a config-only option for testing will I think be a big help in driving maturity and readying this for prime time.

**Comment by BenTheElder** [comment]
> We can save the image size if we just add an installer script in the image.

Kind is generally expected to work fully offline / airgapped.
I'm not sure how helpful this would be to many users versus just bundling it.

> Totally ack this, but to take a devil's advocate position here: Yes, KinD is larger, but as estargz proliferates the net download users should see would be dramatically smaller.

That's not my understanding at all. My understanding is that estargz enables lazy pulling *at the cost* of potentially _larger_ images, but improved latency because of lazy pulling. And of course the kind image will be larger and downloading it is often a major blocking bottleneck in CI flows, so we have to be careful about shipping nifty things that are largely unused / non essential unless we can keep them tiny. (See: rootless)


> So while I don't think we should rush this (certainly as a default), providing a config-only option for testing will I think be a big help in driving maturity and readying this for prime time.

This is available now by using the config from the stargz repo, they provide a kind config + custom image for this.

**Comment by ktock** [comment]
Aiming to make this happen, we started discussion about changes necessary on Stargz Snapshotter's side at https://github.com/containerd/stargz-snapshotter/issues/253.

**Comment by ktock** [comment]
Opened a PR on https://github.com/kubernetes-sigs/kind/pull/2076 .

**Comment by BenTheElder** [comment]
see: https://github.com/kubernetes-sigs/kind/pull/2076#issuecomment-1203047712

TLDR: we're not moving forward with shipping non-standard Kubernetes extensions for now.
